qutebrowser/browser/webkit/network/pac.py

```diff
@@ -0,0 +1,299 @@
+        global_js_object = self._engine.globalObject()
+        self._resolver = global_js_object.property("FindProxyForURL")
+        if not self._resolver.isCallable():
+            err = "Cannot resolve FindProxyForURL function, got '{}' instead"
+            raise EvalProxyError(err.format(self._resolver.toString()))
+
+    def resolve(self, query):
+        """Resolve a proxy via PAC.
+
+        Args:
+            query: QNetworkProxyQuery.
+
+        Return:
+            A list of QNetworkProxy objects in order of preference.
+        """
+        result = self._resolver.call([query.url().toString(),
+                                      query.peerHostName()])
+        result_str = result.toString()
+        if not result.isString():
+            err = "Got strange value from FindProxyForURL: '{}'"
+            raise EvalProxyError(err.format(result_str))
+        return self._parse_proxy_string(result_str)
+
+
+class PACFetcher(QObject):
+
+    """Asynchronous fetcher of PAC files."""
+
+    finished = pyqtSignal()
+
+    def __init__(self, url, parent=None):
+        """Resolve a PAC proxy from URL.
+
+        Args:
+            url: QUrl of a PAC proxy.
+        """
+        super().__init__(parent)
+
+        pac_prefix = "pac+"
+
+        assert url.scheme().startswith(pac_prefix)
+        url.setScheme(url.scheme()[len(pac_prefix):])
+
+        self._manager = QNetworkAccessManager()
+        self._manager.setProxy(QNetworkProxy(QNetworkProxy.NoProxy))
+        self._reply = self._manager.get(QNetworkRequest(url))
+        self._reply.finished.connect(self._finish)
+        self._pac = None
+        self._error_message = None
+
+    @pyqtSlot()
+    def _finish(self):
+        if self._reply.error() != QNetworkReply.NoError:
+            error = "Can't fetch PAC file from URL, error code {}: {}"
+            self._error_message = error.format(self._reply.error(),
+                self._reply.errorString())
+            log.network.error(self._error_message)
+        else:
+            try:
+                pacscript = bytes(self._reply.readAll()).decode("utf-8")
+            except UnicodeError:
+                self._error_message = "Invalid encoding of a PAC file"
+                log.network.exception(self._error_message)
+            try:
+                self._pac = PACResolver(pacscript)
+                log.network.debug("Successfully evaluated PAC file.")
+            except EvalProxyError as e:
+                error = "Error in PAC evaluation: {}"
+                self._error_message = error.format(e)
+                log.network.exception(self._error_message)
+        self._manager = None
+        self._reply = None
+        self.finished.emit()
+
+    def _wait(self):
+        """Wait until a reply from the remote server is received."""
+        if self._manager is not None:
+            loop = qtutils.EventLoop()
+            # Loop exit is triggered on destruction of the reply object.
```

Comment:
Is this still true? I'd say the exit is triggered when `_finish` emits `finished` - even though that's after `self._reply` is destructed, it's not the cause of the loop exit.